### PR TITLE
Add `BufRead::skip_until`

### PR DESCRIPTION
**Pull request by WilliamVenner**
Alternative version of `BufRead::read_until` that simply discards data, rather than copying it into a buffer.

Useful for situations like skipping irrelevant data in a binary file format that is NUL-terminated.

<details>
<summary>Benchmark</summary>

```
running 2 tests
test bench_read_until ... bench:         123 ns/iter (+/- 6)
test bench_skip_until ... bench:          66 ns/iter (+/- 3)
```

```rs
#![feature(test)]
extern crate test;
use test::Bencher;

use std::io::{ErrorKind, BufRead};

fn skip_until<R: BufRead + ?Sized>(r: &mut R, delim: u8) -> Result<usize, std::io::Error> {
    let mut read = 0;
    loop {
        let (done, used) = {
            let available = match r.fill_buf() {
                Ok(n) => n,
                Err(ref e) if e.kind() == ErrorKind::Interrupted => continue,
                Err(e) => return Err(e),
            };
            match memchr::memchr(delim, available) {
                Some(i) => (true, i + 1),
                None => (false, available.len()),
            }
        };
        r.consume(used);
        read += used;
        if done || used == 0 {
            return Ok(read);
        }
    }
}

const STR: &[u8] = b"Ferris\0Hello, world!\0";

#[bench]
fn bench_skip_until(b: &mut Bencher) {
    b.iter(|| {
        let mut io = std::io::Cursor::new(test::black_box(STR));
        skip_until(&mut io, b'\0').unwrap();
        let mut hello = Vec::with_capacity(b"Hello, world!\0".len());
        let num_bytes = io.read_until(b'\0', &mut hello).unwrap();
        assert_eq!(num_bytes, b"Hello, world!\0".len());
        assert_eq!(hello, b"Hello, world!\0");
    });
}


#[bench]
fn bench_read_until(b: &mut Bencher) {
    b.iter(|| {
        let mut io = std::io::Cursor::new(test::black_box(STR));
        io.read_until(b'\0', &mut Vec::new()).unwrap();
        let mut hello = Vec::with_capacity(b"Hello, world!\0".len());
        let num_bytes = io.read_until(b'\0', &mut hello).unwrap();
        assert_eq!(num_bytes, b"Hello, world!\0".len());
        assert_eq!(hello, b"Hello, world!\0");
    });
}
```
</details>

### CODE REVIEW

**Comment by thomcc** [comment]
Hi, thanks for the PR.

As the bot mentioned, you should file an API change proposal for this. This should be done here: https://github.com/rust-lang/libs-team/issues/new?assignees=&labels=api-change-proposal%2C+T-libs-api&template=api-change-proposal.md&title=%28My+API+Change+Proposal%29

See also: <https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html>, which describes the process.

**Comment by pitaj** [comment]
ACP here: https://github.com/rust-lang/libs-team/issues/63

Do create a tracking issue though.

**Comment by thomcc** [comment]
Can you add some tests for this?

@rustbot author

**Comment by WilliamVenner** [comment]
OK - added a test. There's also the doc comment test which is already present.

Is there any reason for the `#[cfg_attr(target_os = "emscripten", ignore)]` on the adjacent `read_until` test? Will that need to be added to the `skip_until` test too? 🤔

File: library/std/src/io/tests.rs [hunk 1]
```diff
@@ -25,6 +25,34 @@ fn read_until() {
...
+        let read = reader.read_until(0, &mut out).unwrap();
+        if read == 0 {
+            // eof
+            break;
+        } else {
+            assert_eq!(out, b"read\0");
+        }
+
+        // skip past `ignore\0`
+        reader.skip_until(0).unwrap();
```
Comment (thomcc) [inline library/std/src/io/tests.rs:47]: Can you assert that `skip_until` is returning the right length?

**Review by thomcc — DISMISSED** [review]
One nit. Otherwise looks fine.

**Review by dtolnay — APPROVED** [review]
Thank you!

**Pull request merged**